internal/image/image_installer.go

```diff
@@ -0,0 +1,152 @@
+package image
+
+import (
+	"fmt"
+	"math/rand"
+	"strconv"
+
+	"github.com/osbuild/osbuild-composer/internal/artifact"
+	"github.com/osbuild/osbuild-composer/internal/common"
+	"github.com/osbuild/osbuild-composer/internal/disk"
+	"github.com/osbuild/osbuild-composer/internal/environment"
+	"github.com/osbuild/osbuild-composer/internal/manifest"
+	"github.com/osbuild/osbuild-composer/internal/platform"
+	"github.com/osbuild/osbuild-composer/internal/rpmmd"
+	"github.com/osbuild/osbuild-composer/internal/runner"
+	"github.com/osbuild/osbuild-composer/internal/users"
+	"github.com/osbuild/osbuild-composer/internal/workload"
+)
+
+type ImageInstaller struct {
+	Base
+	Platform         platform.Platform
+	OSCustomizations manifest.OSCustomizations
+	Environment      environment.Environment
+	Workload         workload.Workload
+
+	ExtraBasePackages rpmmd.PackageSet
+	Users             []users.User
+	Groups            []users.Group
+
+	ISOLabelTempl string
+	Product       string
+	Variant       string
+	OSName        string
+	OSVersion     string
+	Release       string
+
+	Filename string
+}
+
+func NewImageInstaller() *ImageInstaller {
+	return &ImageInstaller{
+		Base: NewBase("image-installer"),
+	}
+}
+
+func (img *ImageInstaller) InstantiateManifest(m *manifest.Manifest,
+	repos []rpmmd.RepoConfig,
+	runner runner.Runner,
+	rng *rand.Rand) (*artifact.Artifact, error) {
+	buildPipeline := manifest.NewBuild(m, runner, repos)
+	buildPipeline.Checkpoint()
+
+	version, err := strconv.Atoi(img.OSVersion)
+
+	if err != nil {
+		panic("cannot convert version to int: " + err.Error())
+	}
+
+	useWebUi := img.OSName == "fedora" && version >= 38
```

Comment:
This is the kind of thing we generally want to avoid outside of the actual distro code.  Basically anything that's in the `image` or `manifest` packages should be distro agnostic.  The distro code itself (in `distro/fedora/`) should understand distribution and version requirements and set options on the pipeline generation code accordingly.

In this case, the way I would do it would be to add a public field to the `ImageInstaller` struct called `AdditionalAnacondaModules` (or just `AnacondaModules`) so that any caller of `NewImageInstaller()` can attach extra modules based on the caller's knowledge of distro version requirements.

There's a similar situation with RHSM stages and options in the code I'm working on for RHEL: https://github.com/osbuild/osbuild-composer/pull/3120/commits/0fa1bca92e8cff0d1e263b37957ac625c8bf90e5

Before, we had RHSM options on the image config for the image types in `distro/rhel9/`, but those options would only be applied (i.e., the rhsm stage would only be added to the pipelines) if the distro was RHEL (i.e., not CentOS).  With the new separation of logic we have now though, the decision moves up into the distro code, so that the RHSM options are conditionally added to the image config and then the pipeline generation code that lives outside `distro/` adds RHSM stages if it's in the image config without needing to check any other conditions.

Reply:
I've rewritten this to be placed into the `distro/fedora/images.go` now by adding `AdditionalKernelOpts` and `AdditionalAnacondaModules` on the `image_installer.go` struct(s).